kong/router.lua

```diff
@@ -1631,180 +1789,32 @@ function _M.new(routes, cache, cache_neg)
-                    else -- ends with / and strip_uri = true, no double slash
-                      upstream_uri = upstream_base .. request_postfix
-                    end
-
-                  else -- ends with / and strip_uri = false
-                    -- we retain the incoming path, just prefix it with the upstream
-                    -- path, but skip the initial slash
-                    upstream_uri = upstream_base .. sub(req_uri, 2)
-                  end
-
-                else -- does not end with /
-                  -- does not end with / and strip_uri = true
-                  if matched_route.strip_uri then
-                    if request_postfix == "" then
-                      if #req_uri > 1 and byte(req_uri, -1) == SLASH then
-                        upstream_uri = upstream_base .. "/"
-                      else
-                        upstream_uri = upstream_base
-                      end
-                    elseif byte(request_postfix, 1, 1) == SLASH then
-                      upstream_uri = upstream_base .. request_postfix
-                    else
-                      upstream_uri = upstream_base .. "/" .. request_postfix
-                    end
-
-                  else -- does not end with / and strip_uri = false
-                    if req_uri == "/" then
-                      upstream_uri = upstream_base
-                    else
-                      upstream_uri = upstream_base .. req_uri
-                    end
-                  end
-                end
-              end
-
-              -- preserve_host header logic
-
-              if matched_route.preserve_host then
-                upstream_host = raw_req_host
-              end
-            end
-
-            local match_t     = {
-              route           = matched_route.route,
-              service         = matched_route.service,
-              headers         = matched_route.headers,
-              upstream_url_t  = upstream_url_t,
-              upstream_scheme = upstream_url_t.scheme,
-              upstream_uri    = upstream_uri,
-              upstream_host   = upstream_host,
-              prefix          = request_prefix,
-              matches         = {
-                uri_captures  = matches.uri_captures,
-                uri           = matches.uri,
-                host          = matches.host,
-                headers       = matches.headers,
-                method        = matches.method,
-                src_ip        = matches.src_ip,
-                src_port      = matches.src_port,
-                dst_ip        = matches.dst_ip,
-                dst_port      = matches.dst_port,
-                sni           = matches.sni,
-              }
-            }
-
-            cache:set(cache_key, match_t)
-
-            return match_t
-          end
-        end
-
-        -- check lower category
-        category_idx = category_idx + 1
+      local match_t = find_match({
```

Comment:
I don't think this extra function adds any clarity (the code is in fact a bit larger after the refactor, given the extra lines needed to set up `ctx`) and, and I am concerned it makes things slower:
* The new `ctx` is bigger, so it takes more cycles to initialize it
* Variables that previously were available as locals must now be accessed via `ctx` (`categories` vs `ctx.categories`)